### PR TITLE
Workaround for DateTimeOffset in  indexes

### LINKED ISSUE #8242: GraphQL + OpenID Error after update to newest version
Hi, after updating OC, there is a failure when generating token by /connect/token endpoint.

System.InvalidCastException: Failed to convert parameter value from a DateTimeOffset to a DateTime.
 ---> System.InvalidCastException: Object must implement IConvertible.
   at System.Convert.ChangeType(Object value, Type conversionType, IFormatProvider provider)
   at Microsoft.Data.SqlClient.SqlParameter.CoerceValue(Object value, MetaType destinationType, Boolean& coercedToDataFeed, Boolean& typeChanged, Boolean allowStreaming)
   --- End of inner exception stack trace ---
   at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__188_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Dapper.SqlMapper.ExecuteWrappedReaderImplAsync(IDbConnection cnn, CommandDefinition command, CommandBehavior commandBehavior) in /_/Dapper/SqlMapper.Async.cs:line 1135
   at YesSql.Commands.BatchCommand.ExecuteAsync(DbConnection connection, DbTransaction transaction, ISqlDialect dialect, ILogger logger)
   at YesSql.Session.FlushAsync()
   at YesSql.Session.CommitAsync()
   at OrchardCore.OpenId.YesSql.Stores.OpenIdTokenStore`1.CreateAsync(TToken token, CancellationToken cancellationToken)
   at OpenIddict.Core.OpenIddictTokenManager`1.CreateAsync(TToken token, CancellationToken cancellationToken)
   at OpenIddict.Core.OpenIddictTokenManager`1.CreateAsync(OpenIddictTokenDescriptor descriptor, CancellationToken cancellationToken)
   at OpenIddict.Core.OpenIddictTokenManager`1.OpenIddict.Abstractions.IOpenIddictTokenManager.CreateAsync(OpenIddictTokenDescriptor descriptor, CancellationToken cancellationToken)
   at OpenIddict.Server.OpenIddictServerHandlers.CreateAccessTokenEntry.HandleAsync(ProcessSignInContext context)
   at OpenIddict.Server.OpenIddictServerDispatcher.DispatchAsync[TContext](TContext context)
   at OpenIddict.Server.OpenIddictServerDispatcher.DispatchAsync[TContext](TContext context)
   at OpenIddict.Server.AspNetCore.OpenIddictServerAspNetCoreHandler.SignInAsync(ClaimsPrincipal user, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Authentication.AuthenticationService.SignInAsync(HttpContext context, String scheme, ClaimsPrincipal principal, AuthenticationProperties properties)
   at Microsoft.AspNetCore.Mvc.SignInResult.ExecuteResultAsync(ActionContext context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|29_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeResultFilters>g__Awaited|27_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddleware.Invoke(HttpContext context)
   at OrchardCore.Apis.GraphQL.GraphQLMiddleware.Invoke(HttpContext context, IAuthorizationService authorizationService, IAuthenticationService authenticationService, ISchemaFactory schemaService)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
   at OrchardCore.Diagnostics.DiagnosticsStartupFilter.<>c__DisplayClass3_0.<<Configure>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Diagnostics.StatusCodePagesMiddleware.Invoke(HttpContext context)
   at OrchardCore.ContentPreview.PreviewStartupFilter.<>c.<<Configure>b__1_1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext)
   at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute, Boolean activateShell)
   at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

### PR DESCRIPTION
**Pull request by jtkech**
Fixes #8242 

Same kind of workaround as done in #8222 

But for now, i was not able to configure openId so that the related tables are populated for testing ;)

@MikeKry would be good if you can try this branch to see if it fixes your issue

**Update**: Also @kevinchalet suggested to also update the type used in the related entities

> If we're fine adopting a breaking change at this point, moving to DateTime shouldn't be a problem and may help a bit with compat' as it has better support outside the SQL Server world.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.OpenId.Core/YesSql/Indexes/OpenIdAuthorizationIndex.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@ public class OpenIdAuthorizationIndex : MapIndex
     {
         public string AuthorizationId { get; set; }
         public string ApplicationId { get; set; }
-        public DateTimeOffset? CreationDate { get; set; }
+        public DateTime? CreationDate { get; set; }
```
Comment (Skrypt) [inline src/OrchardCore/OrchardCore.OpenId.Core/YesSql/Indexes/OpenIdAuthorizationIndex.cs:11]: Here for example if we'd want to keep the offset part we could use a new param/column to store simply the Offset. Should this be needed. Generally for these things a DateTime.UtcDateTime is enough.

Comment (kevinchalet) [reply src/OrchardCore/OrchardCore.OpenId.Core/YesSql/Indexes/OpenIdAuthorizationIndex.cs:11]: Yeah, all these dates are used by OpenIddict to ensure things like tokens are not expired. Loosing the offset info is not a big deal here (and that's what the OpenIddict EF 6/EF Core and MongoDB stores do anyway).

**Comment by jtkech** [comment]
About index tables, i did some tests with sqlite and sql server

If we create an index table with a column of type `DateTimeOffset`, and a `MapIndex` with a property of the same type, on database side it creates a column of type `DateTime`, so we can then change the type in the `MapIndex` from `DateTimeOffset` to `DateTime` without having to do a migration step. I did the test on the `ContentItemIndex`

So seems to be safe to replace `DateTimeOffset` to `DateTime` in migrations / MapIndex

**Comment by kevinchalet** [comment]
> So seems to be safe to replace `DateTimeOffset` to `DateTime` in migrations / MapIndex

Excellent! 🎉 

Note: I see you used `?.UtcDateTime` in the authorization store, but not in the token store. Any particular reason?

**Comment by jtkech** [comment]
> I see you used ?.UtcDateTime in the authorization store, but not in the token store. Any particular reason?

I did it but it doesn't show up in the changed files, you have to expand down the file

**Comment by kevinchalet** [comment]
@jtkech actually, it looks like the token store is already using `?.UtcDateTime` (which is not super logical in the current state as it does `DateTimeOffset` -> `DateTime` -> `DateTimeOffset`) 😅 

Looks like this PR is ready, thanks for fixing that! 👏🏻

**Review by kevinchalet — APPROVED** [review]

**Comment by sebastienros** [comment]
Recommendation is to always store DateTime (not DateTimeOffset) in UTC. 
Only use DateTimeOffset in the public API, if it's not stored.

**Review by sebastienros — APPROVED** [review]

**Comment by jtkech** [comment]
@kevinchalet 

Oops yes it was laready there i didn't change it, i was going crazy to retrieve my update ;)

**Comment by kevinchalet** [comment]
> Recommendation is to always store DateTime (not DateTimeOffset) in UTC.
> Only use DateTimeOffset in the public API, if it's not stored.

And was that a reason to break multiple modules with your YesSql update? :trollface:

**Comment by kevinchalet** [comment]
@jtkech just to be sure, can you check in the logs that you're not seeing weird errors due to `DateTime`/`DateTimeOffset` comparisons when the OpenID background task kicks in? In the EF stores, we had to add `.UtcDateTime`:
  - https://github.com/openiddict/openiddict-core/blob/dev/src/OpenIddict.EntityFrameworkCore/Stores/OpenIddictEntityFrameworkCoreAuthorizationStore.cs#L716
  - https://github.com/openiddict/openiddict-core/blob/dev/src/OpenIddict.EntityFrameworkCore/Stores/OpenIddictEntityFrameworkCoreTokenStore.cs#L686

**Comment by jtkech** [comment]
@kevinchalet 

Yes i saw this but i thought it was okay as now we always set `CreationDate` as an utc time, in case this is the `CreationDate` that is converted to a `DateTimeOffset` for the comparison

Anyway, now better to do it, i will update it

**Comment by MikeKry** [comment]
@jtkech 

Is it safe to try (reversible), or should I create db backup for safety?

**Comment by jtkech** [comment]
@MikeKry 

> Is it safe to try (reversible), or should I create db backup for safety?

Always better to do a backup ;)

**Comment by Kees-Schouten** [comment]
Thanks. This fixed it for me!

**Comment by jtkech** [comment]
@Kees-Schouten Good to know! Thanks

**Pull request merged**